### PR TITLE
Clean up decision_proceduret API

### PR DESCRIPTION
**Pull request by kroening**
This is follow up for #4497; see roadmap there.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/solvers/prop/cover_goals.cpp [hunk 1]
```diff
@@ -48,13 +48,10 @@ void cover_goalst::constraint()
...
-      g_it=goals.begin();
-      g_it!=goals.end();
-      g_it++)
-    if(g_it->status==goalt::statust::UNKNOWN &&
-       !g_it->condition.is_false())
-      disjuncts.push_back(literal_exprt(g_it->condition));
+  for(const auto &g : goals)
+    if(g.status==goalt::statust::UNKNOWN &&
+       !g.condition.is_false())
+      disjuncts.push_back(literal_exprt(g.condition));
```
Comment (tautschnig) [inline src/solvers/prop/cover_goals.cpp]: I believe the next commit reformats this, could this commit already use the correct formatting?

File: src/solvers/prop/cover_goals.h [hunk 2]
```diff
@@ -71,10 +73,9 @@ class cover_goalst
 
   // managing the goals
 
-  void add(const literalt condition)
+  void add(exprt condition)
   {
-    goals.push_back(goalt());
-    goals.back().condition=condition;
+    goals.emplace_back(condition);
```
Comment (tautschnig) [inline src/solvers/prop/cover_goals.h]: I believe this requires `std::move(condition)` to make use of the pass-by-value. (Else I'd suggest to pass a const reference.)

**Review by tautschnig — APPROVED** [review]

File: src/solvers/prop/cover_goals.h (hunk 2 above)
Comment (kroening) [reply src/solvers/prop/cover_goals.h]: ok

File: src/solvers/prop/cover_goals.cpp (hunk 1 above)
Comment (kroening) [reply src/solvers/prop/cover_goals.cpp]: done

File: jbmc/src/jbmc/all_properties.cpp [hunk 3]
```diff
@@ -105,8 +105,7 @@ safety_checkert::resultt bmc_all_propertiest::operator()()
   {
     // Our goal is to falsify a property, i.e., we will
     // add the negation of the property as goal.
-    literalt p = !solver.convert(g.second.as_expr());
-    cover_goals.add(p);
+    cover_goals.add(g.second.as_expr());
```
Comment (peterschrammel) [inline jbmc/src/jbmc/all_properties.cpp]: Where did the negation go?

Comment (kroening) [reply jbmc/src/jbmc/all_properties.cpp]: fixed!

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b885981a6053f96c0712d042dc525cd17cc838f0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108256619

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**